### Loading Data

In [14]:
import pandas as pd
import re

In [15]:
mergedData = '..\\data\\mergedData\\merged_df.csv'
df = pd.read_csv(mergedData)
df.head()

C:\Users\leand\AppData\Local\Temp\ipykernel_13960\1160978504.py:2: DtypeWarning: Columns (1,8,11,14,17,22) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(mergedData)


Unnamed: 0  Report_No Reported_Date Reported_Time   From_Date From_Time  \
0           0  150015427    03/06/2015          0:02  03/06/2015      0:02   
1           3  150057463    09/08/2015          1:17  09/08/2015      1:17   
2           4  150033873    05/19/2015          0:21  05/19/2015      0:21   
3           5  150061779    08/31/2015         10:28  08/31/2015      8:00   
4           6  150087307    12/04/2015          2:57  12/04/2015      2:57   

  To_Date To_Time Offense IBRS  ... Zip Code Rep_Dist Area DVFlag Involvement  \
0     NaN     NaN    2601  90Z  ...    64131   PJ3229  CPD      U         VIC   
1     NaN     NaN     702  240  ...    64126   PJ7474  EPD      U         SUS   
2     NaN     NaN    1849  35A  ...    64128   PJ2340  EPD      U         VIC   
3     NaN     NaN     801  13B  ...    61109   PJ1326  CPD      N         SUS   
4     NaN     NaN    2601  90Z  ...    64108   PJ1326  CPD      U         VIC   

  Race  Sex   Age Firearm Used Flag                  Location  
0  NaN  NaN   NaN                 N  POINT (-94.5767 38.9767)  
1    U    U   NaN                 N  POINT (-94.5516 39.0947)  
2  NaN  NaN   NaN                 N  POINT (-94.5461 39.0735)  
3    B    M  29.0                 N  POINT (-89.0251 42.2167)  
4  NaN  NaN   NaN                 N  POINT (-94.5645 39.0952)  

[5 rows x 24 columns]

### Missing Values

In [16]:
def missingValues(df):
    merged_df_missing_values = df.isnull().sum()
    print(merged_df_missing_values, len(df))

missingValues(df)

Unnamed: 0                0
Report_No                 0
Reported_Date             0
Reported_Time             0
From_Date               871
From_Time              1257
To_Date              592353
To_Time              632103
Offense                   0
IBRS                  53468
Description           68129
Beat                   3448
Address                   0
City                     34
Zip Code              42296
Rep_Dist              95680
Area                   4136
DVFlag                    0
Involvement               0
Race                 131326
Sex                  125034
Age                  350197
Firearm Used Flag         0
Location                  0
dtype: int64 988465


In [17]:
df["Beat"].head()

0    133.0
1    321.0
2    331.0
3    123.0
4    123.0
Name: Beat, dtype: object

In [18]:
df.drop(['To_Time', 'Area', 'Zip Code', 'Beat', 'To_Date', 'From_Time'], axis=1, inplace=True)
missingValues(df)

Unnamed: 0                0
Report_No                 0
Reported_Date             0
Reported_Time             0
From_Date               871
Offense                   0
IBRS                  53468
Description           68129
Address                   0
City                     34
Rep_Dist              95680
DVFlag                    0
Involvement               0
Race                 131326
Sex                  125034
Age                  350197
Firearm Used Flag         0
Location                  0
dtype: int64 988465


In [19]:
df = df.dropna(subset=['From_Date'])
missingValues(df)

Unnamed: 0                0
Report_No                 0
Reported_Date             0
Reported_Time             0
From_Date                 0
Offense                   0
IBRS                  53461
Description           68089
Address                   0
City                     34
Rep_Dist              95672
DVFlag                    0
Involvement               0
Race                 131165
Sex                  124873
Age                  349856
Firearm Used Flag         0
Location                  0
dtype: int64 987594


In [20]:
df['Description'].head(), df['Offense'].head()

(0          Misc Violation
 1              Auto Theft
 2    Possession/Sale/Dist
 3    Non Aggravated Assau
 4          Misc Violation
 Name: Description, dtype: object,
 0    2601
 1     702
 2    1849
 3     801
 4    2601
 Name: Offense, dtype: object)

In [21]:
offense_description_mapping = dict(zip(df['Offense'], df['Description']))
offense_description_mapping

{2601: 'Misc Violation',
 702: 'Auto Theft',
 1849: 'Possession/Sale/Dist',
 801: 'Non Aggravated Assau',
 401: 'Aggravated Assault (',
 201: 'Rape',
 640: 'Stealing From Auto',
 802: 'Non Agg Assault Dome',
 690: 'Stealing All Other',
 1546: 'Weapons Law Violatio',
 302: 'Armed Robbery',
 301: 'Strong Arm Robbery',
 2201: 'Liquor Law Violaton',
 1401: 'Property Damage',
 1651: 'Prostitution/Solicit',
 2402: 'Disorderly Conduct',
 2100: 'Driving Under Influe',
 1652: 'Prostitution/Patroni',
 2403: 'Drunkenness',
 1352: 'Stolen Property OFFE',
 2604: 'False Information to',
 1850: 'Possession of Drug E',
 650: 'Stealing Auto Parts/',
 840: 'Resisting an Officer',
 2040: 'Violation Full Order',
 501: 'Burglary - Residence',
 670: 'Stealing from Buildi',
 402: 'Agg Assault - Drive-',
 2655: 'Trespassing',
 502: 'Burglary - Non Resid',
 799: 'Auto Theft Outside S',
 403: 'Agg Assault - Domest',
 610: 'Stealing Pickpocket',
 2661: 'Kidnapping/Abduction',
 1170: 'Failure to Return',
 2001: '

In [22]:
df[df['Description'].isnull()]

Unnamed: 0   Report_No Reported_Date Reported_Time   From_Date  \
472143           6  KC19022118    03/27/2019          6:23  03/26/2019   
472146          25  KC19030837    04/27/2019         15:14  04/27/2019   
472147          28  KC19033502    05/07/2019         15:07  05/07/2019   
472158         149  KC19027136    04/14/2019          8:50  04/14/2019   
472159         155  KC19023653    04/01/2019         20:30  04/01/2019   
...            ...         ...           ...           ...         ...   
988448       80162  KC24602927    04/25/2024         09:20  04/25/2024   
988454       80168  KC24604676    07/03/2024         05:09  07/01/2024   
988458       80172  KC24600736    01/28/2024         20:09  01/28/2024   
988460       80174  KC24603824    06/03/2024         13:32  06/03/2024   
988462       80176  KC24072383    09/22/2024         15:33  09/22/2024   

                                       Offense IBRS Description  \
472143                               Unfounded  NaN         NaN   
472146             Miscellaneous Investigation  NaN         NaN   
472147             Miscellaneous Investigation  NaN         NaN   
472158            Outside State Warrant Arrest  NaN         NaN   
472159                  Robbery (Armed Street)  NaN         NaN   
...                                        ...  ...         ...   
988448  OCR - Vehicular Non-injury Hit and Run  NaN         NaN   
988454  OCR - Vehicular Non-injury Hit and Run  NaN         NaN   
988458  OCR - Vehicular Non-injury Hit and Run  NaN         NaN   
988460  OCR - Vehicular Non-injury Hit and Run  NaN         NaN   
988462                    City Weapons Offense  NaN         NaN   

                                Address         City Rep_Dist DVFlag  \
472143                3800  NW BARRY RD  KANSAS CITY      NaN      N   
472146                1600  TOPPING AVE  KANSAS CITY   PJ1559      N   
472147                 4500  MERCIER ST  KANSAS CITY      NaN      N   
472158                 1600  GRAND BLVD  KANSAS CITY   PJ1399      N   
472159                  3500  E 59TH ST  KANSAS CITY   PJ4681      N   
...                                 ...          ...      ...    ...   
988448                3800  NW 85TH TER  KANSAS CITY   PP0324  False   
988454                 2700  MCGEE TRFY  KANSAS CITY   PJ2192  False   
988458  N BRIGHTON AVE and N I 35 HWY N  KANSAS CITY   PC0960  False   
988460             8800  N CONGRESS AVE  KANSAS CITY   PP0287  False   
988462                 7500  WORNALL RD  KANSAS CITY   PJ5638  False   

        Involvement Race  Sex   Age Firearm Used Flag  \
472143      CMP VIC    B    M  51.0             False   
472146  ARR CHA SUS    W    F  37.0             False   
472147          VIC    W    M  34.0             False   
472158          VIC  NaN  NaN   NaN             False   
472159          SUS    B    M   NaN             False   
...             ...  ...  ...   ...               ...   
988448          SUS    U    U   NaN             False   
988454      CMP VIC    W    F  24.0             False   
988458      CMP VIC    W    F  23.0             False   
988460      CMP VIC    W    F  28.0             False   
988462      CMP VIC    B    M  23.0             False   

                                  Location  
472143        POINT (-94.624219 39.246545)  
472146         POINT (-94.514818 39.09262)  
472147        POINT (-94.601519 39.045569)  
472158        POINT (-94.581186 39.094363)  
472159        POINT (-94.545285 39.018933)  
...                                    ...  
988448  POINT (-94.626121068 39.251030231)  
988454   POINT (-94.58087276 39.077996474)  
988458  POINT (-94.521313698 39.183452115)  
988460  POINT (-94.668092573 39.255583732)  
988462  POINT (-94.594658033 38.991870014)  

[68089 rows x 18 columns]

In [23]:
df[df['Description'].isnull()]['Offense'].apply(lambda x: isinstance(x, str)).all()

True

Description           68089
Leandro Die wer sind nur null wil sie händ en System änderig gmacht wo die Offence mal number frühner gsi ish denn händ in description gschribä was es ish. Aber jetzt machets das nümme die schriebet alles in Offence ine darum sind die description null. Als Note

In [24]:
# Check the type of each value in the 'Offense' column
offense_types = df['Offense'].apply(type)

# Count the number of string and numeric values
num_strings = (offense_types == str).sum()
num_numbers = (offense_types == int).sum()

print(f"Number of string values in 'Offense': {num_strings}")
print(f"Number of numeric values in 'Offense': {num_numbers}")

Number of string values in 'Offense': 529652
Number of numeric values in 'Offense': 457942


In [26]:
# Function to check if a value can be converted to int
def can_convert_to_int(value):
    try:
        string_value = int(value)
        return string_value.is_integer()  # Check if the string is an integer
    except (ValueError, TypeError):
        return False

# Apply the conversion check and count valid rows
valid_offense_count = df['Offense'].apply(can_convert_to_int).sum()

# Output the number of rows where 'Offense' can be converted to int
print(f"Number of rows where 'Offense' can be converted to int: {valid_offense_count}")

Number of rows where 'Offense' can be converted to int: 471315


In [27]:
# Filter rows where 'Offense' is numeric
numeric_offense_df = df[df['Offense'].apply(lambda x: can_convert_to_int(x))]

# Check if all numeric 'Offense' values have a corresponding 'Description' that is a string
all_numeric_offense_have_string_description = numeric_offense_df['Description'].apply(lambda x: isinstance(x, str)).all()

print(f"All numeric 'Offense' values have a string 'Description': {all_numeric_offense_have_string_description}")

All numeric 'Offense' values have a string 'Description': False


In [28]:
df[df['Offense'].apply(lambda x: can_convert_to_int(x))]

Unnamed: 0  Report_No Reported_Date Reported_Time   From_Date Offense  \
0                0  150015427    03/06/2015          0:02  03/06/2015    2601   
1                3  150057463    09/08/2015          1:17  09/08/2015     702   
2                4  150033873    05/19/2015          0:21  05/19/2015    1849   
3                5  150061779    08/31/2015         10:28  08/31/2015     801   
4                6  150087307    12/04/2015          2:57  12/04/2015    2601   
...            ...        ...           ...           ...         ...     ...   
472140      128935  180098804    12/24/2018         22:39  12/24/2018     831   
472141      128936  180099422    12/27/2018         21:29  12/27/2018    1401   
472142      128937  180099911    12/30/2018         05:00  12/30/2018     401   
530920       76230  190004971    01/21/2019          8:22  01/19/2019       5   
533341       79471  190004971    01/21/2019          8:22  01/19/2019       5   

       IBRS           Description                      Address         City  \
0       90Z        Misc Violation     BROADWAY and WESTPORT RD  KANSAS CITY   
1       240            Auto Theft  PROSPECT AV and E TRUMAN RD  KANSAS CITY   
2       35A  Possession/Sale/Dist     VICTOR ST and WALROND AV  KANSAS CITY   
3       13B  Non Aggravated Assau        PASEO and E TRUMAN RD  KANSAS CITY   
4       90Z        Misc Violation        PASEO and E TRUMAN RD  KANSAS CITY   
...     ...                   ...                          ...          ...   
472140  13C          Intimidation        5600  N MANCHESTER AV  KANSAS CITY   
472141  290       Property Damage              7400  WABASH AV  KANSAS CITY   
472142  13A  Aggravated Assault (           7000  CLEVELAND AV  KANSAS CITY   
530920  35A                   NaN          9701 MARION PARK DR  KANSAS CITY   
533341  35A                   NaN          9701 MARION PARK DR  KANSAS CITY   

       Rep_Dist DVFlag Involvement Race  Sex   Age Firearm Used Flag  \
0        PJ3229      U         VIC  NaN  NaN   NaN                 N   
1        PJ7474      U         SUS    U    U   NaN                 N   
2        PJ2340      U         VIC  NaN  NaN   NaN                 N   
3        PJ1326      N         SUS    B    M  29.0                 N   
4        PJ1326      U         VIC  NaN  NaN   NaN                 N   
...         ...    ...         ...  ...  ...   ...               ...   
472140   PC0692      N         VIC    B    M  29.0                 N   
472141   PJ5630      U         SUS    U    U   NaN                 N   
472142   PJ5365      N         VIC    B    F  26.0                 Y   
530920   PJ6588      U         VIC  NaN  NaN   NaN             False   
533341   PJ6588      U         SUS    B    F  36.0             False   

                            Location  
0           POINT (-94.5767 38.9767)  
1           POINT (-94.5516 39.0947)  
2           POINT (-94.5461 39.0735)  
3           POINT (-89.0251 42.2167)  
4           POINT (-94.5645 39.0952)  
...                              ...  
472140   POINT (-94.48833 39.195536)  
472141  POINT (-94.558376 38.992484)  
472142  POINT (-94.542832 38.999755)  
530920   POINT (-94.53572 38.948543)  
533341   POINT (-94.53572 38.948543)  

[471315 rows x 18 columns]

In [13]:
df['Offense_Description'] = df.apply(
    lambda row: row['Offense'] if isinstance(row['Offense'], str) else row['Description'], axis=1
)

df[['Offense', 'Description', 'Offense_Description']].head()

Offense           Description   Offense_Description
0    2601        Misc Violation        Misc Violation
1     702            Auto Theft            Auto Theft
2    1849  Possession/Sale/Dist  Possession/Sale/Dist
3     801  Non Aggravated Assau  Non Aggravated Assau
4    2601        Misc Violation        Misc Violation

In [14]:
offense_description_types = df['Offense_Description'].apply(type)
offense_description_types.value_counts(), df.__len__()

(Offense_Description
 <class 'str'>    987594
 Name: count, dtype: int64,
 987594)

In [15]:
df.drop(['Offense', 'Description'], axis=1, inplace=True)
missingValues(df)

Unnamed: 0                  0
Report_No                   0
Reported_Date               0
Reported_Time               0
From_Date                   0
IBRS                    53461
Address                     0
City                       34
Rep_Dist                95672
DVFlag                      0
Involvement                 0
Race                   131165
Sex                    124873
Age                    349856
Firearm Used Flag           0
Location                    0
Offense_Description         0
dtype: int64 987594


In [16]:
mode_sex = df['Sex'].mode()[0]

df['Sex'].fillna(mode_sex, inplace=True)

print(df['Sex'].isnull().sum())

0


C:\Users\leand\AppData\Local\Temp\ipykernel_9996\1862509000.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Sex'].fillna(mode_sex, inplace=True)


In [17]:
missingValues(df)

Unnamed: 0                  0
Report_No                   0
Reported_Date               0
Reported_Time               0
From_Date                   0
IBRS                    53461
Address                     0
City                       34
Rep_Dist                95672
DVFlag                      0
Involvement                 0
Race                   131165
Sex                         0
Age                    349856
Firearm Used Flag           0
Location                    0
Offense_Description         0
dtype: int64 987594


In [18]:
df = df.dropna(subset=['City'])
missingValues(df)

Unnamed: 0                  0
Report_No                   0
Reported_Date               0
Reported_Time               0
From_Date                   0
IBRS                    53444
Address                     0
City                        0
Rep_Dist                95651
DVFlag                      0
Involvement                 0
Race                   131156
Sex                         0
Age                    349842
Firearm Used Flag           0
Location                    0
Offense_Description         0
dtype: int64 987560


In [19]:
df.drop(['Rep_Dist', 'IBRS', 'Age', 'Race', 'Sex'], axis=1, inplace=True)
missingValues(df)

Unnamed: 0             0
Report_No              0
Reported_Date          0
Reported_Time          0
From_Date              0
Address                0
City                   0
DVFlag                 0
Involvement            0
Firearm Used Flag      0
Location               0
Offense_Description    0
dtype: int64 987560


### Export

In [20]:
df.to_csv("..\\data\\mergedData\\merged_df.csv")